types/wlr_screencopy_v1.c

```diff
@@ -269,18 +282,39 @@ static void frame_handle_copy(struct wl_client *wl_client,
 		return;
 	}
 
-	struct wl_shm_buffer *buffer = wl_shm_buffer_get(buffer_resource);
-	if (buffer == NULL) {
+	struct wlr_dmabuf_v1_buffer* dma_buffer = NULL;
+	struct wl_shm_buffer *shm_buffer = wl_shm_buffer_get(buffer_resource);
+	if (shm_buffer == NULL) {
+		dma_buffer =
+			wlr_dmabuf_v1_buffer_from_buffer_resource(buffer_resource);
+	}
+
+	if (shm_buffer == NULL && dma_buffer == NULL ) {
 		wl_resource_post_error(frame->resource,
 			ZWLR_SCREENCOPY_FRAME_V1_ERROR_INVALID_BUFFER,
 			"unsupported buffer type");
 		return;
 	}
 
-	enum wl_shm_format fmt = wl_shm_buffer_get_format(buffer);
-	int32_t width = wl_shm_buffer_get_width(buffer);
-	int32_t height = wl_shm_buffer_get_height(buffer);
-	int32_t stride = wl_shm_buffer_get_stride(buffer);
+	enum wl_shm_format fmt = 0;
+	int32_t width = 0;
+	int32_t height = 0;
+	int32_t stride = 0;
+
+	if (shm_buffer) {
+		fmt = wl_shm_buffer_get_format(shm_buffer);
+		width = wl_shm_buffer_get_width(shm_buffer);
+		height = wl_shm_buffer_get_height(shm_buffer);
+		stride = wl_shm_buffer_get_stride(shm_buffer);
+	} else if(dma_buffer) {
+		width = dma_buffer->attributes.width;
+		height = dma_buffer->attributes.height;
+
+		// Doesn't matter for dmabuf
+		fmt = frame->format;
```

Comment:
Should we allow the client to choose an arbitrary format here? We don't want the compositor to do arbitrary format conversions.

Reply:
Doesn't the format conversion happen on the GPU when it's sampling the input texture and writing it into the dmabuf, given that the GPU supports the output format?

This assignment is perhaps a bit misleading, I did this to make the `fmt == frame->format` check below always true for the dmabuf case.

Reply:
The format conversion may be costly.

Reply:
I don't think we should be allowing clients to choose formats, at least not without providing them with a hint and some info on what conversions are allowed. I'd still rather not let them choose their formats.
In Vulkan image copies and downloads are generally restricted to only allow same formats.

Reply:
Unless we want to make a backwards incompatible change to the protocol the format still needs to be `wl_shm_format`, even for dmabuf. Is that ok?

Reply:
I think that's the idea.

Reply:
Well, I pushed two commits to that effect yesterday, so the format is enforced now.